[Lock] DynamoDB store

natepage opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #59996  <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

This PR introduces a new Lock store using AWS DynamoDb. The idea is coming from this [article](https://aws.amazon.com/blogs/database/building-distributed-locks-with-the-dynamodb-lock-client/).

The implementation is heavily based on:

- [DoctrineDbalStore](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Lock/Store/DoctrineDbalStore.php) for the overall logic
- [AmazonSqsMessenger Connection](https://github.com/symfony/amazon-sqs-messenger/blob/7.3/Transport/Connection.php) for the DSN/options handling

natepage committed:
[Lock] DynamoDB store

stof commented:
this new store should probably in a bridge package, so that we can manage its dependency properly.

natepage commented:
@stof Thank you for your feedback, I did move it to its own bridge, I'm not sure if I've done everything required from a structural POV.

I still didn't touch the tests yet

OskarStark commented on src/Symfony/Component/Lock/Bridge/DynamoDb/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+7.3
+---
+
+* Introduced the Amazon DynamoDb bridge.
```
```suggestion
* Add the bridge
```

OskarStark commented on src/Symfony/Component/Lock/Bridge/DynamoDb/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2020-present Fabien Potencier
```
```suggestion
Copyright (c) 2025-present Fabien Potencier
```

OskarStark commented on src/Symfony/Component/Lock/Bridge/DynamoDb/README.md:
```diff
@@ -0,0 +1,12 @@
+Amazon DynamoDb Lock
```
```suggestion
Amazon DynamoDB Lock
```

OskarStark commented on src/Symfony/Component/Lock/Bridge/DynamoDb/README.md:
```diff
@@ -0,0 +1,12 @@
+Amazon DynamoDb Lock
+====================
+
+Provides Amazon DynamoDb integration for Symfony Lock.
```
```suggestion
Provides Amazon DynamoDB integration for Symfony Lock.
```
Should be spelled like this, only in namespaces or classnames we should use DyanomDb with a lowercased b

OskarStark commented:
> this new store should probably in a bridge package, so that we can manage its dependency properly.

While I agree on that @stof, right now there is no Bridge namespace:
![CleanShot 2025-04-09 at 12 14 02@2x](https://github.com/user-attachments/assets/07a2ca59-f4ff-42fc-9973-661ac9b4c7b5)

stof commented:
@OskarStark this does not prevent us from _adding_ it if we need it.

natepage commented:
@stof @OskarStark thank you for taking the time to review this 😄 

I've actually updated the PR to add this Bridge namespace following the structure from the Messenger SQS example, but the package validation is failing and I'm unsure why as I've placed the composer.json file exactly in the same level as the Messenger SQS one, trying to investigate that right now

natepage commented:
Yay! Package checks are passing, the match case for `component_bridge` wasn't behaving as expected, back onto PHP tests 😄

OskarStark commented:
Yes, we need to keep in mind, that we need to configure a proper split for the repo etc.

stof commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,277 @@
+        'access_key' => null,
+        'secret_key' => null,
+        'session_token' => null,
+        'endpoint' => 'https://dynamodb.us-west-1.amazonaws.com',
```
The default should not configure an endpoint for the async-aws client but should let it use the default endpoint for the appropriate region (this endpoint value is specific to a region)

natepage replied:
@stof thank you for that, I've updated the default options to be nullable and given to the async-aws client only when set. I've applied this to region as well as otherwise the following DSN `dynamodb://default/lock_keys` would have defaulted to the hardcoded region instead of the region resolved from the async-aws client.

natepage commented:
@stof @OskarStark

The current state of this is:

- Moved the logic into its own component bridge
- Implemented integration tests by using localstack in CI to have a local DynamoDB instance
- Updated the default options logic based on feedback
- Implemented unit tests for the store constructor logic (DSN/Options)

I have re-requested your reviews, thank you 😄

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Tests/Functional/Store/AbstractStoreTestCase.php:
```diff
@@ -0,0 +1,126 @@
+/**
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+abstract class AbstractStoreTestCase extends TestCase
```
Instead of duplicating this base test case, the class should be moved to `Symfony\Component\Lock\Test\StoreTestCase` (`Test` without `s`), so that the class is exposed by the `symfony/lock` package, and becomes part of its API. Then it can be used by all the bridge packages.

natepage replied:
@GromNaN I have to say I didn't feel good copy/pasting that class across, but wasn't sure exposing it as part of the package was something we would want. But great news, makes things much cleaner!

I see in your comment you renamed it to `StoreTestCase`, is that intentional or could we keep it to `AbstractStoreTestCase`?

GromNaN replied:
We don't prefix the shared TestCase with `Abstract` in [other components](https://github.com/search?q=repo%3Asymfony%2Fsymfony+path%3A**%2F*TestCase.php&type=code). You can remove it.

natepage replied:
Done

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/composer.json:
```diff
@@ -0,0 +1,40 @@
+        "async-aws/core": "^1.7",
+        "async-aws/dynamo-db": "^3.0",
+        "symfony/lock": "^7.3",
+        "symfony/service-contracts": "^2.5|^3",
+        "psr/log": "^1|^2|^3"
+    },
+    "require-dev": {
+        "symfony/polyfill-uuid": "^1.15",
+        "symfony/http-client-contracts": "^2.5|^3"
+    },
+    "conflict": {
+        "symfony/http-client-contracts": "<2.5"
+    },
```
Looking at the `DynamoDbStore` class, I don't get why `symfony/service-contracts`, `psr/log`, `symfony/polyfill-uuid` and `symfony/http-client-contracts` need to be explicitly listed as dependencies.

natepage replied:
@GromNaN that's a miss from when I copied the files structure from the Lock component. 
I've removed most of the packages you listed, however I did explicitly add `symfony/polyfill-uuid` to `^1.15` as otherwise it would default to `1.13` and this version is failing as it's looking for a constant that doesn't exist and was breaking the CI.

Maybe I'm missing something and there is a better way here?

stof replied:
What error do you get exactly ?

Is this `async-aws/dynamodb` using a uuid feature that is not yet polyfilled in older version of the polyfill package, or is it the polyfill package itself being broken ?

GromNaN replied:
The constants were not correctly declared in 1.13.0: ([diff](https://github.com/symfony/polyfill-uuid/compare/v1.13.0...v1.14.0#diff-d11e2b9e464a2d6711842693ee58d997e7d238e803fbbe33cea332ece518159eL15-L25)). This blocks PHPUnit execution.

```
PHPUnit 9.6.22 by Sebastian Bergmann and contributors.

Error in bootstrap script: Error:
Undefined constant "UUID_VARIANT_NCS"
#0 src/Symfony/Component/Lock/Bridge/DynamoDb/vendor/composer/autoload_real.php(41): require()
```

natepage replied:
@GromNaN beat me to it, was sleeping on this side of the world 😄 

You can see this failure in this [run](https://github.com/symfony/symfony/actions/runs/14350059000/job/40226931778#step:9:10669)

@stof Thanks for that, and for the PR in async-aws

GromNaN commented on src/Symfony/Component/Lock/Store/StoreFactory.php:
```diff
@@ -27,6 +29,9 @@ class StoreFactory
     {
         switch (true) {
+            case $connection instanceof DynamoDbClient:
+                return new DynamoDbStore($connection);
```
It would be nice to add a specific error message when the package is missing. I suggest adding a private function to this class.

```php
private static function requireBridgeClass(string $class, string $package): void
{
    if (!class_exists($class)) {
        throw new \LogicException(sprintf('Class "%s" is missing. Try running "composer require %s" to install the bridge.), $class, $package);
    }
}
```

And then:
```suggestion
                self::requireBridgeClass(DynamoDbStore::class, 'symfony/amazon-dynamodb-lock');

                return new DynamoDbStore($connection);
```

natepage replied:
Done

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/composer.json:
```diff
@@ -0,0 +1,34 @@
+        "symfony/lock": "^7.3"
+    },
+    "require-dev": {
+        "symfony/polyfill-uuid": "^1.15"
```
> I did explicitly add symfony/polyfill-uuid to ^1.15 as otherwise it would default to 1.13 and this version is failing as it's looking for a constant that doesn't exist and was breaking the CI.

`symfony/polyfill-uuid` is a dependency of `async-aws/dynamo-db`, if it requires a newer version, [its `composer.json`](https://github.com/async-aws/dynamo-db/blob/3a2c922bffb7b30edb64c0522a49f39a3a9255fe/composer.json#L18) should be updated.

Since UUID are not directly used in this bridge package, I would use a "conflict" rule. So that we don't require it if the async-aws package switches to something else.

natepage replied:
Done

GromNaN commented on src/Symfony/Component/Lock/Store/StoreFactory.php:
```diff
@@ -60,6 +67,9 @@ public static function createStore(#[\SensitiveParameter] object|string $connect
                 return new SemaphoreStore();
 
+            case str_starts_with($connection, 'dynamodb://'):
+                return new DynamoDbStore($connection);
```
This class check is also necessary here.
```suggestion
                self::requireBridgeClass(DynamoDbStore::class, 'symfony/amazon-dynamodb-lock');

                return new DynamoDbStore($connection);
```

natepage replied:
Done

GromNaN commented on .github/get-modified-packages.php:
```diff
@@ -22,7 +22,7 @@ function getPackageType(string $packageDir): string
         str_contains($packageDir, 'Symfony/Bridge/') => 'bridge',
         str_contains($packageDir, 'Symfony/Bundle/') => 'bundle',
-        preg_match('@Symfony/Component/[^/]+/Bridge/@', $packageDir) => 'component_bridge',
+        1 === preg_match('@Symfony/Component/[^/]+/Bridge/@', $packageDir) => 'component_bridge',
```
> Yay! Package checks are passing, the match case for component_bridge wasn't behaving as expected, back onto PHP tests 😄

Thank you for pointing that out. 

I opened a PR to backport this fix into 7.2: https://github.com/symfony/symfony/pull/60216

GromNaN reviewed (commented):
Just a few more adjustments for this new package, and it looks good to me. I haven't read over the implementation in detail.

GromNaN commented on src/Symfony/Component/Lock/Store/StoreFactory.php:
```diff
@@ -115,4 +127,11 @@ public static function createStore(#[\SensitiveParameter] object|string $connect
+    private static function requireBridgeClass(string $class, string $package): void
+    {
+        if (!class_exists($class)) {
+            throw new \LogicException(\sprintf('Class "%s" is missing. Try running "%s" to install the bridge.', $class, 'composer require '.$package));
```
```suggestion
            throw new \LogicException(\sprintf('Class "%s" is missing. Try running "composer require %s" to install the lock store package.', $class, $package));
```

natepage replied:
@GromNaN I've made that change to keep fabbot happy.

I've implemented your suggested change here, and as you can see [there](https://fabbot.io/report/symfony/symfony/60138/a0739ce9e42cee40777bcfa3e1f906def6033504), fabbot requires to wrap each parameter with double quotes, which in this case would result in "double double quotes" 😄 

If we are ok with the path proposed by fabbot, then I will go ahead and implement it but it looked strange to me, please let me know.

GromNaN replied:
fabbot isn't always right; we sometimes ignore fixes like this one.

GromNaN commented on src/Symfony/Component/Lock/Tests/Store/RedisStoreTestCase.php:
```diff
@@ -18,11 +18,12 @@
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-abstract class AbstractRedisStoreTestCase extends AbstractStoreTestCase
+abstract class RedisStoreTestCase extends StoreTestCase
```
Renaming this class wasn't necessary, but it's consistent.

natepage replied:
Yes that was a mass refactor miss, reverted it back.

fabpot replied:
```suggestion
abstract class AbstractRedisStoreTestCase extends AbstractStoreTestCase
```

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,277 @@
+        if ($clientOrUrl instanceof DynamoDbClient) {
+            $this->client = $clientOrUrl;
+        } else {
+            if (!class_exists(DynamoDbClient::class)) {
+                throw new InvalidArgumentException(\sprintf('You cannot use the "%s" if the DynamoDbClient is not available. Try running "composer require async-aws/dynamo-db".', __CLASS__));
+            }
+
```
This is not necessary, since the class has its dedicated package, the `async-aws/dynamo-db` package is required by composer.

```suggestion
```

natepage replied:
Done

GromNaN reviewed (approved):
SGTM.
Just a line of code that can be removed.

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/README.md:
```diff
@@ -0,0 +1,12 @@
+Amazon DynamoDB Lock
+====================
+
+Provides Amazon DynamoDB integration for Symfony Lock.
```
We can add a small doc for the DSN to use, as we do for some bridges ([example for Brevo](https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Notifier/Bridge/Brevo/README.md)).

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,284 @@
+            }
+
+            // check for extra keys in options
+            $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
```
Nitpick: `array_diff_key` would be more appropriate.

GromNaN commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,284 @@
+            }
+            unset($query['region']);
+
+            if ('default' !== ($params['host'] ?? 'default')) {
+                $clientConfiguration['endpoint'] = \sprintf('%s://%s%s', ($options['sslmode'] ?? null) === 'disable' ? 'http' : 'https', $params['host'], ($params['port'] ?? null) ? ':'.$params['port'] : '');
+                if (preg_match(';^dynamodb\.([^\.]++)\.amazonaws\.com$;', $params['host'], $matches)) {
+                    $clientConfiguration['region'] = $matches[1];
+                }
```
I don't see where this code is tested. Can you add a test case with a custom host?

natepage replied:
@GromNaN This [test](https://github.com/symfony/symfony/pull/60138/files#diff-ad209a8e7643dd458b2823bf4401a13a8b72dfb08bc09519cba84e1e2ba4cfbeR130) is setting `host => localhost` which tests that part

natepage commented:
Hi there! 👋 
I'm sure everybody is quite busy prepping for 7.3 release, but just a gentle follow up on this one, 
I would like to know if anything else is needed from me so I can organise myself to have time to spend on this, thank you 😄

stof commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,284 @@
+    private int $writeCapacityUnits;
+
+    public function __construct(
+        DynamoDbClient|string $clientOrUrl,
```
I would name it `$clientOrDsn` instead

natepage replied:
@stof thank you, done

fabpot commented on src/Symfony/Component/Lock/Bridge/DynamoDb/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+7.3
```
```suggestion
7.4
```

fabpot commented on src/Symfony/Component/Lock/Test/StoreTestCase.php:
```diff
@@ -19,7 +19,7 @@
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-abstract class AbstractStoreTestCase extends TestCase
+abstract class StoreTestCase extends TestCase
```
This change should be reverted as we always prefix abstract classes with `Abstract`.

fabpot commented on src/Symfony/Component/Lock/Bridge/DynamoDb/Store/DynamoDbStore.php:
```diff
@@ -0,0 +1,284 @@
+    public function __construct(
+        DynamoDbClient|string $clientOrDsn,
+        array $options = [],
+        private readonly int $initialTtl = 300
```
```suggestion
        private readonly int $initialTtl = 300,
```
